token-server/src/AuthClient.ts

```diff
@@ -0,0 +1,118 @@
+import { AccessToken } from "@bentley/itwin-client";
+import { decode } from "jsonwebtoken";
+import {
+  GrantBody,
+  TokenSet,
+  ClientMetadata,
+  Issuer,
+  Client as OpenIdClient,
+} from "openid-client";
+
+/**
+ * Utility to generate OIDC/OAuth tokens for agent or agent applications
+ * * The application must register a client using the
+ * [self service registration page](https://developer.bentley.com/register).
+ * * The client type must be "Service"
+ * * Use the Client Id/Client Secret/Scopes to create the client.
+ * * Ensure the application can access the iTwin Project/Asset - in production environments, this is done by
+ * using the iTwin project portal to add add the email **`{Client Id}@apps.imsoidc.bentley.com`** as an authorized user
+ * with the appropriate role that includes the required access permissions.
+ */
+export class AuthClient {
+  private _client?: OpenIdClient;
+  private _accessToken?: AccessToken;
+
+  constructor(
+    private _clientId: string,
+    private _clientSecret: string,
+    private _scopes: string
+  ) {}
+
+  private async _getClient(): Promise<OpenIdClient> {
+    if (this._client) {
+      return this._client;
+    }
+
+    const clientConfiguration: ClientMetadata = {
+      client_id: this._clientId,
+      client_secret: this._clientSecret, // eslint-disable-line @typescript-eslint/naming-convention
+    };
+    const issuer = await Issuer.discover("https://ims.bentley.com");
+    this._client = new issuer.Client(clientConfiguration);
+    return this._client;
+  }
+
+  private async _generateAccessToken(): Promise<AccessToken> {
+    const scope = this._scopes;
```

Comment:
return cached access token, also probably want to invalidate it/set it to undefined when it does expire, unless we can renew a token before it expires

Reply:
This should only be called when I'm not authorized, which checks to see if the access token already exists so should be good... Unless I'm missing something here.